File: src/temporal_xray/transformers/diff_engine.py
from __future__ import annotations

import json
from typing import Any

from ..types import ExecutionDivergence, TimelineStep, WorkflowHistory


def diff_executions(
    history_a: WorkflowHistory,
    history_b: WorkflowHistory,
) -> dict[str, Any]:
    """Compare two workflow execution histories and identify divergences."""
    divergences = _find_data_divergences(history_a.timeline, history_b.timeline)
    structural = _find_structural_differences(history_a.timeline, history_b.timeline)
    signals = _find_signal_differences(history_a, history_b)
    return {
        "divergences": divergences,
        "structural_differences": structural,
        "signals": signals,
    }


def _find_data_divergences(
    timeline_a: list[TimelineStep],
    timeline_b: list[TimelineStep],
) -> list[ExecutionDivergence]:
    divergences: list[ExecutionDivergence] = []
    aligned = _align_timelines(timeline_a, timeline_b)

    for pair_a, pair_b in aligned:
        if pair_a is None or pair_b is None:
            continue

        input_a = pair_a.input if pair_a.input is not None else pair_a.input_summary
        input_b = pair_b.input if pair_b.input is not None else pair_b.input_summary
        for diff in _deep_diff(input_a, input_b, "input"):
            divergences.append(ExecutionDivergence(
                step=pair_a.step,
                activity=pair_a.activity,
                field=diff["path"],
                value_a=diff["value_a"],
                value_b=diff["value_b"],
                note=diff["note"],
            ))

        output_a = pair_a.output if pair_a.output is not None else pair_a.output_summary
        output_b = pair_b.output if pair_b.output is not None else pair_b.output_summary
        for diff in _deep_diff(output_a, output_b, "output"):
            divergences.append(ExecutionDivergence(
                step=pair_a.step,
                activity=pair_a.activity,
                field=diff["path"],
                value_a=diff["value_a"],
                value_b=diff["value_b"],
                note=diff["note"],
            ))

        if pair_a.status != pair_b.status:
            divergences.append(ExecutionDivergence(
                step=pair_a.step,
                activity=pair_a.activity,
                field="status",
                value_a=pair_a.status,
                value_b=pair_b.status,
                note=f"Activity {pair_a.activity} has different status in each execution",
            ))

        if (pair_a.retries or 0) != (pair_b.retries or 0):
            divergences.append(ExecutionDivergence(
                step=pair_a.step,
                activity=pair_a.activity,
                field="retries",
                value_a=pair_a.retries or 0,
                value_b=pair_b.retries or 0,
                note=f"Different retry counts for {pair_a.activity}",
            ))

    return divergences


def _find_structural_differences(
    timeline_a: list[TimelineStep],
    timeline_b: list[TimelineStep],
) -> dict[str, Any]:
    types_a = [s.activity for s in timeline_a]
    types_b = [s.activity for s in timeline_b]

    set_a = set(types_a)
    set_b = set(types_b)

    only_in_a = list({t for t in types_a if t not in set_b})
    only_in_b = list({t for t in types_b if t not in set_a})

    common_a = [t for t in types_a if t in set_b]
    common_b = [t for t in types_b if t in set_a]
    different_order = (
        len(common_a) > 0
        and len(common_b) > 0
        and common_a != common_b
    )

    return {
        "activities_only_in_a": only_in_a,
        "activities_only_in_b": only_in_b,
        "different_execution_order": different_order,
    }


def _find_signal_differences(
    history_a: WorkflowHistory,
    history_b: WorkflowHistory,
) -> dict[str, list[str]]:
    sig_names_a = {s["name"] for s in history_a.signals_received}
    sig_names_b = {s["name"] for s in history_b.signals_received}
    return {
        "signals_only_in_a": [s for s in sig_names_a if s not in sig_names_b],
        "signals_only_in_b": [s for s in sig_names_b if s not in sig_names_a],
    }


def _align_timelines(
    timeline_a: list[TimelineStep],
    timeline_b: list[TimelineStep],
) -> list[tuple[TimelineStep | None, TimelineStep | None]]:
    pairs: list[tuple[TimelineStep | None, TimelineStep | None]] = []
    used_b: set[int] = set()

    for step_a in timeline_a:
        matched = False
        for i, step_b in enumerate(timeline_b):
            if i in used_b:
                continue
            if step_b.activity == step_a.activity:
                pairs.append((step_a, step_b))
                used_b.add(i)
                matched = True
                break
        if not matched:
            pairs.append((step_a, None))

    for i, step_b in enumerate(timeline_b):
        if i not in used_b:
            pairs.append((None, step_b))

    return pairs


def _deep_diff(a: Any, b: Any, base_path: str) -> list[dict[str, Any]]:
    if a is b or a == b:
        return []

    if (
        isinstance(a, dict)
        and isinstance(b, dict)
    ):
        results: list[dict[str, Any]] = []
        all_keys = set(a.keys()) | set(b.keys())
        for key in all_keys:
            results.extend(_deep_diff(a.get(key), b.get(key), f"{base_path}.{key}"))
        return results

    if isinstance(a, list) and isinstance(b, list):
        results = []
        max_len = max(len(a), len(b))
        for i in range(max_len):
            val_a = a[i] if i < len(a) else None
            val_b = b[i] if i < len(b) else None
            results.extend(_deep_diff(val_a, val_b, f"{base_path}[{i}]"))
        return results

    return [{
        "path": base_path,
        "value_a": a,
        "value_b": b,
        "note": f"Different values at {base_path}",
    }]

File: src/transformers/diff-engine.ts
import type { TimelineStep, ExecutionDivergence, WorkflowHistory } from "../types.js";

/**
 * Compare two workflow execution histories and identify divergences.
 */
export function diffExecutions(
  historyA: WorkflowHistory,
  historyB: WorkflowHistory
): {
  divergences: ExecutionDivergence[];
  structural_differences: {
    activities_only_in_a: string[];
    activities_only_in_b: string[];
    different_execution_order: boolean;
  };
  signals: {
    signals_only_in_a: string[];
    signals_only_in_b: string[];
  };
} {
  const divergences = findDataDivergences(historyA.timeline, historyB.timeline);
  const structural = findStructuralDifferences(historyA.timeline, historyB.timeline);
  const signals = findSignalDifferences(historyA, historyB);

  return { divergences, structural_differences: structural, signals };
}

/**
 * Find data divergences: same activity at same step, different inputs/outputs.
 */
function findDataDivergences(
  timelineA: TimelineStep[],
  timelineB: TimelineStep[]
): ExecutionDivergence[] {
  const divergences: ExecutionDivergence[] = [];

  // Align activities by matching type and sequence
  const aligned = alignTimelines(timelineA, timelineB);

  for (const pair of aligned) {
    if (!pair.a || !pair.b) continue;

    // Compare inputs
    const inputA = pair.a.input ?? pair.a.input_summary;
    const inputB = pair.b.input ?? pair.b.input_summary;
    const inputDiffs = deepDiff(inputA, inputB, `input`);

    for (const diff of inputDiffs) {
      divergences.push({
        step: pair.a.step,
        activity: pair.a.activity,
        field: diff.path,
        value_a: diff.valueA,
        value_b: diff.valueB,
        note: diff.note,
      });
    }

    // Compare outputs
    const outputA = pair.a.output ?? pair.a.output_summary;
    const outputB = pair.b.output ?? pair.b.output_summary;
    const outputDiffs = deepDiff(outputA, outputB, `output`);

    for (const diff of outputDiffs) {
      divergences.push({
        step: pair.a.step,
        activity: pair.a.activity,
        field: diff.path,
        value_a: diff.valueA,
        value_b: diff.valueB,
        note: diff.note,
      });
    }

    // Compare status
    if (pair.a.status !== pair.b.status) {
      divergences.push({
        step: pair.a.step,
        activity: pair.a.activity,
        field: "status",
        value_a: pair.a.status,
        value_b: pair.b.status,
        note: `Activity ${pair.a.activity} has different status in each execution`,
      });
    }

    // Compare retries
    if ((pair.a.retries || 0) !== (pair.b.retries || 0)) {
      divergences.push({
        step: pair.a.step,
        activity: pair.a.activity,
        field: "retries",
        value_a: pair.a.retries || 0,
        value_b: pair.b.retries || 0,
        note: `Different retry counts for ${pair.a.activity}`,
      });
    }
  }

  return divergences;
}

/**
 * Find structural differences: activities that exist in one execution but not the other.
 */
function findStructuralDifferences(
  timelineA: TimelineStep[],
  timelineB: TimelineStep[]
): {
  activities_only_in_a: string[];
  activities_only_in_b: string[];
  different_execution_order: boolean;
} {
  const typesA = timelineA.map((s) => s.activity);
  const typesB = timelineB.map((s) => s.activity);

  const setA = new Set(typesA);
  const setB = new Set(typesB);

  const onlyInA = typesA.filter((t) => !setB.has(t));
  const onlyInB = typesB.filter((t) => !setA.has(t));

  // Check if activities common to both were executed in the same order
  const commonA = typesA.filter((t) => setB.has(t));
  const commonB = typesB.filter((t) => setA.has(t));
  const differentOrder =
    commonA.length > 0 &&
    commonB.length > 0 &&
    JSON.stringify(commonA) !== JSON.stringify(commonB);

  return {
    activities_only_in_a: [...new Set(onlyInA)],
    activities_only_in_b: [...new Set(onlyInB)],
    different_execution_order: differentOrder,
  };
}

/**
 * Find signal differences between two executions.
 */
function findSignalDifferences(
  historyA: WorkflowHistory,
  historyB: WorkflowHistory
): {
  signals_only_in_a: string[];
  signals_only_in_b: string[];
} {
  const sigNamesA = new Set(historyA.signals_received.map((s) => s.name));
  const sigNamesB = new Set(historyB.signals_received.map((s) => s.name));

  return {
    signals_only_in_a: [...sigNamesA].filter((s) => !sigNamesB.has(s)),
    signals_only_in_b: [...sigNamesB].filter((s) => !sigNamesA.has(s)),
  };
}

// --- Alignment and Diff Utilities ---

interface AlignedPair {
  a: TimelineStep | null;
  b: TimelineStep | null;
}

/**
 * Align two timelines by activity type and sequence.
 * Uses a simple greedy matching: for each step in A, find the first unmatched
 * step in B with the same activity type.
 */
function alignTimelines(
  timelineA: TimelineStep[],
  timelineB: TimelineStep[]
): AlignedPair[] {
  const pairs: AlignedPair[] = [];
  const usedB = new Set<number>();

  for (const stepA of timelineA) {
    let matched = false;
    for (let i = 0; i < timelineB.length; i++) {
      if (usedB.has(i)) continue;
      if (timelineB[i].activity === stepA.activity) {
        pairs.push({ a: stepA, b: timelineB[i] });
        usedB.add(i);
        matched = true;
        break;
      }
    }
    if (!matched) {
      pairs.push({ a: stepA, b: null });
    }
  }

  // Add unmatched B steps
  for (let i = 0; i < timelineB.length; i++) {
    if (!usedB.has(i)) {
      pairs.push({ a: null, b: timelineB[i] });
    }
  }

  return pairs;
}

interface DiffResult {
  path: string;
  valueA: unknown;
  valueB: unknown;
  note: string;
}

/**
 * Deep diff two values and return a list of field-level differences.
 */
function deepDiff(a: unknown, b: unknown, basePath: string): DiffResult[] {
  if (a === b) return [];
  if (a === undefined && b === undefined) return [];
  if (a === null && b === null) return [];

  // If both are strings (summaries), compare directly
  if (typeof a === "string" && typeof b === "string") {
    if (a !== b) {
      return [
        {
          path: basePath,
          valueA: a,
          valueB: b,
          note: `Different values at ${basePath}`,
        },
      ];
    }
    return [];
  }

  // If both are objects, compare field by field
  if (
    a !== null &&
    b !== null &&
    typeof a === "object" &&
    typeof b === "object" &&
    !Array.isArray(a) &&
    !Array.isArray(b)
  ) {
    const results: DiffResult[] = [];
    const allKeys = new Set([
      ...Object.keys(a as Record<string, unknown>),
      ...Object.keys(b as Record<string, unknown>),
    ]);

    for (const key of allKeys) {
      const valA = (a as Record<string, unknown>)[key];
      const valB = (b as Record<string, unknown>)[key];
      const childPath = `${basePath}.${key}`;
      results.push(...deepDiff(valA, valB, childPath));
    }

    return results;
  }

  // If both are arrays, compare element by element
  if (Array.isArray(a) && Array.isArray(b)) {
    const results: DiffResult[] = [];
    const maxLen = Math.max(a.length, b.length);

    for (let i = 0; i < maxLen; i++) {
      results.push(...deepDiff(a[i], b[i], `${basePath}[${i}]`));
    }

    return results;
  }

  // Different types or primitive values
  if (a !== b) {
    return [
      {
        path: basePath,
        valueA: a,
        valueB: b,
        note: `Different values at ${basePath}`,
      },
    ];
  }

  return [];
}
